### PR TITLE
C#: Introduce inherited CFG completions

### PR DESCRIPTION
**Pull request by hvitved**
When completions are inherited by elements inside `finally` blocks, we previously
threw away the underlying completion. For example, in

```
try
{
    if (b)
        throw new Exception();
}
finally
{
    if (b)
        ...
}
```

the completions for `b` inside the `finally` block are `true` and `throw(Exception)`,
where the latter is inherited from the `try` block, with an underlying `false`
completion. Throwing away the `false` completion meant that we were unable to prune
the `false` edge (Boolean CFG splitting).

Profiling report [here](https://git.semmle.com/gist/tom/f361c2a619cbde3709622c1a13adf227#file-report2-md) (internal link).

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll [hunk 1]
```diff
@@ -668,165 +658,142 @@ abstract private class InheritableCompletion extends AbnormalCompletion { }
  * `System.Console.WriteLine("M called")` inherits the throw completion
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll]: We should remove the word "inherits" and "underlying" and replace it with "has the outer" and "has the inner"

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll [hunk 2]
```diff
@@ -668,165 +658,142 @@ abstract private class InheritableCompletion extends AbnormalCompletion { }
...
-
-private class GotoLabelCompletionInherited extends GotoLabelCompletion, InheritedCompletion {
-  private TGotoLabelCompletion inherited;
+  GotoLabelStmt getGotoStmt() { result = goto }
 
-  GotoLabelCompletionInherited() { this = TInheritedCompletion(_, inherited) }
-
-  override GotoLabelStmt getGotoStmt() { inherited = TGotoLabelCompletion(result) }
+  override string toString() {
+    this = TGotoLabelCompletion(goto) and result = "goto(" + this.getLabel() + ")"
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll]: This is a little sketchy. I suppose, morally, there should be a `or result = this.(NestedCompletion).toString()` to ensure that `toString()` always exists.

Comment (calumgrant) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll]: Ditto for the cases below.

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll]: I know, how about a comment instead?

**Pull request merged**